File: test/AdventureWorksLT/SalesOrderInfoDetail.cs
using DevZest.Data;
using DevZest.Data.Annotations;

namespace DevZest.Samples.AdventureWorksLT
{
    [ExtraColumns(typeof(ForeignKeyLookup.Ext))]
    public class SalesOrderInfoDetail : SalesOrderDetail
    {
        [ModelValidator]
        private DataValidationError ValidateProduct(DataRow dataRow)
        {
            if (ProductID[dataRow] == null)
                return null;
            var ext = GetExtraColumns<ForeignKeyLookup.Ext>();
            var productNumber = ext.Product.ProductNumber;
            var productName = ext.Product.Name;

            if (string.IsNullOrEmpty(productNumber[dataRow]))
                return new DataValidationError(string.Format(UserMessages.Validation_ValueIsRequired, productNumber.DisplayName), productNumber);

            if (string.IsNullOrEmpty(productName[dataRow]))
                return new DataValidationError(string.Format(UserMessages.Validation_ValueIsRequired, productName.DisplayName), productName);

            return null;
        }
    }
}

File: test/AdventureWorksLT.VB/SalesOrderInfoDetail.vb
Imports DevZest.Data
Imports DevZest.Data.Annotations

Namespace DevZest.Samples.AdventureWorksLT
    <ExtraColumns(GetType(Product.Lookup))>
    Public Class SalesOrderInfoDetail
        Inherits SalesOrderDetail

        <ModelValidator>
        Private Function ValidateProduct(ByVal dataRow As DataRow) As DataValidationError
            If ProductID(dataRow) Is Nothing Then Return Nothing
            Dim ext = GetExtraColumns(Of Product.Lookup)()
            Dim productNumber = ext.ProductNumber
            Dim productName = ext.Name
            If String.IsNullOrEmpty(productNumber(dataRow)) Then Return New DataValidationError(String.Format(UserMessages.Validation_ValueIsRequired, productNumber.DisplayName), productNumber)
            If String.IsNullOrEmpty(productName(dataRow)) Then Return New DataValidationError(String.Format(UserMessages.Validation_ValueIsRequired, productName.DisplayName), productName)
            Return Nothing
        End Function
    End Class
End Namespace
